File: Tree/814. Binary Tree Pruning.cpp
/*
814. Binary Tree Pruning

We are given the head node root of a binary tree, where additionally every node's value is either a 0 or a 1.

Return the same tree where every subtree (of the given tree) not containing a 1 has been removed.

(Recall that the subtree of a node X is X, plus every node that is a descendant of X.)

Example 1:
Input: [1,null,0,0,1]
Output: [1,null,0,null,1]
 
Explanation: 
Only the red nodes satisfy the property "every subtree not containing a 1".
The diagram on the right represents the answer.


Example 2:
Input: [1,0,1,0,0,0,1]
Output: [1,null,1,null,1]



Example 3:
Input: [1,1,0,1,1,0,1,0]
Output: [1,1,0,1,1,null,1]



Note:

The binary tree will have at most 100 nodes.
The value of each node will only be 0 or 1.

*/

class Solution {
public:
    TreeNode* pruneTree(TreeNode* root) {
        if(!root) return nullptr;
        root->left = pruneTree(root->left);
        root->right = pruneTree(root->right);
        if(!root->left && !root->right && root->val == 0) return nullptr;
        return root;
    }
};


class Solution {
public:
    TreeNode* pruneTree(TreeNode* root) {
        check(&root);
        return root;
    }
    
    void check(TreeNode** root) {
        if (*root != NULL) {
            TreeNode* temp = *root;
            check(&(temp->left));
            check(&(temp->right));
            if (temp->left == NULL && temp->right == NULL && temp->val == 0) {
                *root = nullptr;
            }
        }
    }
};

class Solution {
public:
    TreeNode* pruneTree(TreeNode* root) {
        if (root) root->left = pruneTree(root->left), root->right = pruneTree(root->right);
        return (root && (root->left || root->right || root->val)) ? root : NULL;
    }
};

File: Tree/814. Binary Tree Pruning.py
"""
814. Binary Tree Pruning


We are given the head node root of a binary tree, where additionally every node's value is either a 0 or a 1.

Return the same tree where every subtree (of the given tree) not containing a 1 has been removed.

(Recall that the subtree of a node X is X, plus every node that is a descendant of X.)

"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def pruneTree(self, root):
        def dfs(root):
            if not root: return None
            root.left = dfs(root.left)
            root.right = dfs(root.right)
            if not root.left and not root.right and not root.val:
                return None
            return root
        return dfs(root)
        

class Solution:
    def pruneTree(self, root):
        dummy = TreeNode(1)
        dummy.left = root
        stack = [(0, dummy)]
        while stack:
            seen, node = stack.pop()
            if node is None:
                continue
            if not seen:
                stack.extend([(1, node), (0, node.right), (0, node.left)])
            else:
                if node.left and not (node.left.val or node.left.left or node.left.right):
                    node.left = None
                if node.right and not (node.right.val or node.right.left or node.right.right):
                    node.right = None
        return dummy.left 


#对于全是0的tree，返回[0] 而我们需要返回[]
class Solution:
    def pruneTree(self, root):
        def dfs(root):
            if not root: return True
            left = dfs(root.left)
            right = dfs(root.right)
            if left: root.left = None
            if right: root.right = None 
            return left and right and root.val != 1
        dfs(root)
        return root
        